### DISCUSSION TITLE
Switch

### DISCUSSION THREAD

**Discussion by Rozerinay** [nl-design-system/backlog#293] [Component suggestions]
<img width="480" alt="visuele weergave van een switch component" src="https://github.com/user-attachments/assets/76d0ef82-ba5c-4b6a-96a7-affc8922ad12">

We willen graag meer leren over de Switch component. In welke vorm, of vormen, wordt deze momenteel ingezet en toegepast binnen jouw organisatie?

Deel hieronder een link naar Figma, Storybook, Github of screenshots. Met deze informatie kunnen we bewijzen dat de component onderdeel moet worden van de NL Design System component library.

Daarnaast wordt het delen van onderzoek gewaardeerd. Dus heb je zelf onderzoek gedaan en zijn er inzichten voor dit component? Dan horen we dit ook graag!

[Link naar bericht in Slack]({link-naar-Slack-bericht})

---

## Naam
Switch

## Doel
Invoerveld om een optie aan of uit te zetten, waarbij de actie direct wordt doorgevoerd.

## Ook bekend als
Toggle, Toggle Switch, Form Toggle

## Gerelateerde componenten
[Checkbox](https://github.com/orgs/nl-design-system/discussions/306), [Radio Button](https://github.com/orgs/nl-design-system/discussions/288)

## Onderzoek

[Toggle-Switch Guidelines - NN Group](https://www.nngroup.com/articles/toggle-switch-guidelines/)
In dit artikel beschrijft Alita Joyce paar richtlijnen voor een Switch component. Ze zet de verschillen tussen Checkboxes en Radio Buttons naast elkaar en geef enkele tips voor het ontwerp van een Switch.

[Toggle button - Inclusive components](https://inclusive-components.design/toggle-button/)
Uitgebreid artikel met afwegingen die komen kijken bij het implementeren van een Toggle button.

[On Designing and Building Toggle Switches - Sara Soueidan](https://www.sarasoueidan.com/blog/toggle-switch-design/)
Uitgebreid artikel met afwegingen die komen kijken bij het implementeren van een Toggle Switch.

[Under-Engineered Toggles - Adrian Roselli](https://adrianroselli.com/2019/03/under-engineered-toggles.html)
In dit artikel beschrijft Ardian Roselli zijn afwegingen bij het ontwikkelen van een Toggle op basis van een **Checkbox**.

[Under-Engineered Toggles - Adrian Roselli](https://adrianroselli.com/2019/08/under-engineered-toggles-too.html)
In dit artikel beschrijft Ardian Roselli zijn afwegingen bij het ontwikkelen van een Toggle op basis van een **Button**.

[Switch Role Support - Adrian Roselli](https://adrianroselli.com/2021/10/switch-role-support.html)
In dit artikel beschrijft Ardian Roselli de support voor de Switch role in combinatie met de Checkbox en Button. En welke invloed dit heeft op schermlezers.

[Toggles suck! - Axess Lab](https://axesslab.com/toggles-suck/)
In dit artikel beschrijft Joel Holmberg waarom hij Toggles verschrikkelijk vindt. Hij geeft hiervoor als reden dat de status van de Toggle vaak niet goed is te begrijpen. Dat ze niet toegankelijk zijn. Dat ze inconsistent worden gebruikt op het web en er geen native HTML ondersteuning is. Hij adviseerd om Checkboxes of Radio Buttons te gebruiken.

[Why Toggle Switches suck, and what to do instead - Adam Silver](https://adamsilver.io/blog/why-toggle-switches-suck-and-what-to-do-instead/)

[Switch in HTML - Hidde de Vries](https://hidde.blog/links/switch-in-html/)
In deze blogpost beschrijft Hidde dat Apple experimenteert met Switch als variant op een Checkbox.

**Comment by jeffreylauwers** [comment]
## Amsterdam

- [Storybook](https://designsystem.amsterdam/?path=/docs/components-forms-switch--docs)
- [Figma](https://www.figma.com/design/9IGm6IdPUYizBNGsUnueBd/Amsterdam-Design-System?node-id=2927-29193)
- [GitHub](https://github.com/Amsterdam/design-system/tree/develop/packages/css/src/components/switch)

---

### Screenshots / Varianten

**Storybook**

Default
![Scherm­afbeelding 2024-10-16 om 14 00 40](https://github.com/user-attachments/assets/4d539136-f18a-4d0f-b343-99ec4f123bf6)

Hover
![Scherm­afbeelding 2024-10-16 om 14 00 52](https://github.com/user-attachments/assets/fbe982bf-0e62-427c-9f81-677bd2461f76)

Focus
![Scherm­afbeelding 2024-10-16 om 14 00 58](https://github.com/user-attachments/assets/6b35a469-097e-4e66-b90f-3a708d57dd66)

Disabled
![Scherm­afbeelding 2024-10-16 om 14 01 30](https://github.com/user-attachments/assets/d31a358c-b017-4b7c-af18-d7fa8ba3b3b7)

Checked
![Scherm­afbeelding 2024-10-16 om 14 01 06](https://github.com/user-attachments/assets/71a45285-05c3-4103-8313-2b0b589ed1f6)

Checked + Hover
![Scherm­afbeelding 2024-10-16 om 14 01 14](https://github.com/user-attachments/assets/f022ace0-1196-4e60-a58c-f841c30590c6)

Checked + Focus
![Scherm­afbeelding 2024-10-16 om 14 01 19](https://github.com/user-attachments/assets/9d32e64b-8fbe-4aec-bf23-b99fa27523ae)

Checked + Disabled
![Scherm­afbeelding 2024-10-16 om 14 01 38](https://github.com/user-attachments/assets/da0997db-ee16-44f4-864d-909d238a78ed)

With label
![Scherm­afbeelding 2024-10-16 om 14 01 48](https://github.com/user-attachments/assets/b996fc56-7daf-4cce-88e2-bf553ab4ca1b)

---

### Notities

- Documentatie beschikbaar.
- Unchecked: Thumb links, grijze achtergrond.
- Checked: Thumb rechts, donderblauwe achtergrond.

**Comment by jeffreylauwers** [comment]
## Den Haag

[Figma - Huidig](https://www.figma.com/design/JpoY3waVoQGlLQzQXTL9nn/%F0%9F%8C%BB-Den-Haag-Design-System?node-id=1681-7439&t=mYo78jne5kP62APW-4)

---

### Screenshots / Varianten

**Figma**

States
![Scherm­afbeelding 2024-10-16 om 14 10 05](https://github.com/user-attachments/assets/602b1e42-e98f-4fc7-a1b6-179db1ec53a7)


---

### Notities

- Onder de naam: Toggle
- Documentatie beschikbaar (Figma)
- Unchecked: Thumb links, grijze achtergrond.
- Checked: Thumb rechts, groene achtergrond.

**Comment by jeffreylauwers** [comment]
## DSO

[Storybook](https://storybook.dso-toolkit.nl/66.0.0/?path=/docs/core-slide-toggle--docs)

---

### Screenshots / Varianten

**Storybook**

Default
![Scherm­afbeelding 2024-10-16 om 14 37 30](https://github.com/user-attachments/assets/7c2a4585-353d-435d-be8d-813dda9915d1)

Focus
![Scherm­afbeelding 2024-10-16 om 14 39 34](https://github.com/user-attachments/assets/1993e7e1-2d1c-41ef-bcdb-43a88f5561a5)

Disabled
![Scherm­afbeelding 2024-10-16 om 14 37 45](https://github.com/user-attachments/assets/54ef4b60-0f2f-4925-bc22-c9cde94d18dc)

Checked
![Scherm­afbeelding 2024-10-16 om 14 37 39](https://github.com/user-attachments/assets/3353249f-fb7b-49c3-b3fd-3a9bd3531e6a)

Checked + Focus
![Scherm­afbeelding 2024-10-16 om 14 39 45](https://github.com/user-attachments/assets/561ff227-6b1c-4c4e-bc4d-cada368f4b26)

Labelled by ID
![Scherm­afbeelding 2024-10-16 om 14 38 02](https://github.com/user-attachments/assets/5e84d9e6-680f-4442-a664-b71d06826800)

Zichtbaar label
![Scherm­afbeelding 2024-10-16 om 14 38 13](https://github.com/user-attachments/assets/e60ae7f6-2386-4d10-81ff-647cc83789d6)

---

### Notities

- Onder de naam: Slide Toggle
- Unchecked: Thumb links, grijze achtergrond.
- Checked: Thumb rechts, donker groene achtergrond.

**Comment by jeffreylauwers** [comment]
## DUO

[Documentatie website](https://uno.dfront.rijkscloud.nl/#/components/switch)

---

### Screenshots / Varianten

**Documentatie website**

Algemeen voorbeeld
![Scherm­afbeelding 2024-10-16 om 14 44 07](https://github.com/user-attachments/assets/542f1f26-9337-46bb-be3d-0fff40c06858)

States unchecked
![Scherm­afbeelding 2024-10-16 om 14 43 21](https://github.com/user-attachments/assets/daaee15f-2328-4408-a64d-4a1f36165469)

States checked
![Scherm­afbeelding 2024-10-16 om 14 43 35](https://github.com/user-attachments/assets/880c0436-17f9-44e6-9ca5-2928c831415b)

Switch in horizontaal: Een horizontale switch
![Scherm­afbeelding 2024-10-16 om 14 43 44](https://github.com/user-attachments/assets/c082a2d9-837c-4fb2-9f19-ac09a4e13f8a)

Table switch: Een switch als deze binnen een tabel wordt geplaatst.
![Scherm­afbeelding 2024-10-16 om 14 43 52](https://github.com/user-attachments/assets/4ddc63df-e121-45b4-a2ef-334c375e38cb)

---

### Notities

- Documentatie beschikbaar
- Unchecked: Thumb links (zwart), witte achtergrond, zwarte 2px solid border. Rechts het woord 'Uit'.
- Checked: Thumb rechts (wit), blauwe achtergrond, border krijgt zelfde kleur als achtergrond. Rechts het woord 'Aan'.
- Varianten verschillen in formaat. Kleiner binnen een tabel.

**Comment by jeffreylauwers** [comment]
## Rotterdam

[Figma](https://www.figma.com/design/ZWSC4gCrOXRUR9UX3aoZ8x/Gemeente-Rotterdam-Design-System?node-id=521-27335&t=YBcq3pZvpnJOckOZ-4)

---

### Screenshots / Varianten

**Figma**

Default
![Scherm­afbeelding 2024-10-16 om 14 53 36](https://github.com/user-attachments/assets/1d58bd70-e58f-4179-aaba-fd995615b8f6)

Focus
![Scherm­afbeelding 2024-10-16 om 14 53 45](https://github.com/user-attachments/assets/9b82374d-e6a5-4af2-8fde-0175e1bf4c50)


Checked
![Scherm­afbeelding 2024-10-16 om 14 53 40](https://github.com/user-attachments/assets/80672abe-b6eb-40e1-8535-fb46595bf3ef)


---

### Notities

- Switch heeft rechte hoeken.
- Thumb is een wit vierkant met 6 lichtgrijze stippen.
- Unchecked: Thumb links, grijze achtergrond.
- Checked: Thumb rechts, groene achtergrond.

**Comment by jeffreylauwers** [comment]
## Utrecht

- [Storybook](https://nl-design-system.github.io/utrecht/storybook/?path=/docs/css_css-form-toggle--docs)
- [GitHub](https://github.com/nl-design-system/utrecht/tree/main/components/form-toggle)

---

### Screenshots / Varianten

**Storybook**

Default
![Scherm­afbeelding 2024-10-16 om 15 12 49](https://github.com/user-attachments/assets/007e617e-835a-4b27-8b16-1a4fd6da6a19)

With invisible `<input type="checkbox">`
![Scherm­afbeelding 2024-10-16 om 15 12 56](https://github.com/user-attachments/assets/efe8b90b-465e-44db-960f-fa129af9331a)

Hover
![Scherm­afbeelding 2024-10-16 om 15 13 03](https://github.com/user-attachments/assets/41d3d90e-5af6-4276-8d83-f8112a389a0d)

Focus
![Scherm­afbeelding 2024-10-16 om 15 13 08](https://github.com/user-attachments/assets/de39d163-6d20-47c3-a4a3-8d0c90fd2579)

Focus-visible
![Scherm­afbeelding 2024-10-16 om 15 13 13](https://github.com/user-attachments/assets/084c4f77-a312-4db4-8f61-b23d40a5123f)

Checked
![Scherm­afbeelding 2024-10-16 om 15 13 19](https://github.com/user-attachments/assets/fae291df-d2ea-4fc7-a573-9ac502a29847)

Disabled
![Scherm­afbeelding 2024-10-16 om 15 13 23](https://github.com/user-attachments/assets/85ffc224-fd22-48f9-a949-73eec396b8a6)

Checked and disabled
![Scherm­afbeelding 2024-10-16 om 15 13 29](https://github.com/user-attachments/assets/214d56c9-eed8-4fcd-a187-2402f4619191)

---

### Notities

- Onder de naam: Form Toggle
- Unchecked: Thumb links, grijze achtergrond.
- Checked: Thumb rechts, donker blauwe achtergrond.

**Reply by Rozerinay** [reply]
## Vragen en overwegingen bij componentontwikkeling

Component is tijdens Estafettemodeldag 2 opgebouwd op basis van de Utrecht component.

Vragen die zijn meegegeven voor verdere uitwerking:

* Thumb: ook een token voor min-block-size toevoegen?
* Margin: doen we niet in Figma, maar zouden we de track een padding kunnen geven?
* Color en background color: waar zijn deze tokens precies voor bedoeld?
* Box shadow: alleen toepassen op de disabled thumb? Indien we een shadow toevoegen, zou dat eerder voor de abled state verwacht worden dan voor de disabled state.
* Focus tokens: waarom nu wel, terwijl dit bijvoorbeeld bij een button niet gebeurt?
* Extra border tokens: toevoegen voor de thumb zodat de vorm zichtbaar blijft in high-contrast mode?
* Icoon toevoegen: zodat er niet alleen op kleur wordt vertrouwd voor de ‘checked/unchecked’ state?

Bron: notities uit een comment in de NL Design System Figma bibliotheek (2024).

**Reply by Rozerinay** [reply]
Opmerking afkomstig uit een comment in de NL Design System Figma bibliotheek:

Na overleg met Robbert icoon toegevoegd.
https://github.com/nl-design-system/kernteam/issues/847

**Comment by MariekeBrouwer** [comment]
![image](https://github.com/user-attachments/assets/7402ea52-9f4e-453b-a19c-b449d554ea5d)
Bij Groningen wordt hij momenteel gebruik voor een dark modus, waarin we eigenlijk ook bij de documentatie erna hebben gedefinieerd dat :
- het alleen gebruikt wordt wanneer er een directe wijziging is in de interface
- bij gebruik van meerdere switches onder elkaar, moeten de opties niet gerelateerd zijn aan elkaar (bv. bluetooth aan/uit. wifi aan/uit)

**Comment by jeffreylauwers** [comment]
Aanvulling van Rian in [draadje Hiep Hoi bericht](https://codefornl.slack.com/archives/C7KDQRXMZ/p1729681202736039).

> Of “waarbij de actie direct wordt doorgevoerd” aan WCAG voldoet verdient nog een discussie.
> https://nldesignsystem.nl/wcag/3.2.2/
> 
> Maar bij een settings pagina zie ik het vaak en ik vind het een mooie oplossing. Een save optie aan het eind is wellicht wel zo fijn.
> To be continued. Dus, ja als component toevoegen, en nadenken over de exacte implementatie.

**Comment by jeffreylauwers** [comment]
## ✨ Dit component is nu Help Wanted ✨

Help je mee hem door de Community stappen te krijgen?
[Switch op het Community bord](https://github.com/orgs/nl-design-system/projects/29/views/1?filterQuery=switch)

**Comment by JJGrootveld** [comment]
[In de ToDo Bibliotheek](https://www.figma.com/design/42KfIBw43pwLxv2tE7oRzg/NL-Design-System---ToDo-Bibliotheek?node-id=1763-592&t=eEIJ428jDjZnZoez-0) staat op dit moment een `disabled` state gedocumenteerd voor de Switch component. 

Wat is de rationale achter het wel aanbieden van deze state t.o.v. het sterke afraden van disabled states [in de Button component](https://www.figma.com/design/FqAr99wvrlHxTJYAHkFRQN/NL-Design-System---Bibliotheek?node-id=11475-470&t=giFLmcaKTxy9EVMO-11)?

Misschien trek ik dit te ver door, maar is een Switch niet eigenlijk ook een submit button voor een specifieke actie? 
https://www.nldesignsystem.nl/richtlijnen/formulieren/buttons/disabled-submitbuttons/#disabled-submitbuttons

Als de actie die gelinkt is aan de toggle niet door de gebruiker uitgevoerd kan worden, moet dat dan uberhaupt getoond worden?

**Comment by eslook** [comment]
## ✨ Dit component is nu Community ✨


Help je mee hem door de Candidate stappen te krijgen?
[Switch op het Candidate bord](https://github.com/orgs/nl-design-system/projects/80/views/1?filterQuery=switch)
